pallets/nft/src/lib.rs

```diff
@@ -0,0 +1,412 @@
+            ticker.clone(),
+            portfolio_kind,
+        )?;
+
+        Portfolio::<T>::ensure_portfolio_validity(&caller_portfolio)?;
+
+        // Verifies that all mandatory keys are being set and that there are no duplicated keys
+        let mandatory_keys: BTreeSet<AssetMetadataKey> = Self::collection_keys(&collection_id);
+        ensure!(
+            mandatory_keys.len() == metadata_attributes.len(),
+            Error::<T>::InvalidMetadataAttribute
+        );
+
+        let n_keys = metadata_attributes.len();
+        let nft_attributes: BTreeMap<_, _> = metadata_attributes
+            .into_iter()
+            .map(|a| (a.key, a.value))
+            .collect();
+        ensure!(
+            n_keys == nft_attributes.len(),
+            Error::<T>::DuplicateMetadataKey
+        );
+
+        for metadata_key in nft_attributes.keys() {
+            ensure!(
+                mandatory_keys.contains(metadata_key),
+                Error::<T>::InvalidMetadataAttribute
+            );
+        }
+
+        // Mints the NFT and adds it to the caller's portfolio
+        let new_balance = BalanceOf::get(&ticker, &caller_portfolio.did)
+            .checked_add(ONE_UNIT)
+            .ok_or(Error::<T>::BalanceOverflow)?;
+        let nft_id = NextNFTId::try_mutate(&collection_id, try_next_pre::<T, _>)?;
+        BalanceOf::insert(&ticker, &caller_portfolio.did, new_balance);
+        for (metadata_key, metadata_value) in nft_attributes.into_iter() {
+            MetadataValue::insert((&collection_id, &nft_id), metadata_key, metadata_value);
+        }
+        PortfolioNFT::insert(caller_portfolio, (ticker, nft_id), true);
+
+        Self::deposit_event(Event::MintedNft(
+            caller_portfolio.did,
+            collection_id,
+            nft_id,
+        ));
+        Ok(())
+    }
+
+    fn base_burn_nft(
+        origin: T::Origin,
+        ticker: Ticker,
+        nft_id: NFTId,
+        portfolio_kind: PortfolioKind,
+    ) -> DispatchResult {
+        // Verifies if the collection exists
+        let collection_id =
+            CollectionTicker::try_get(&ticker).map_err(|_| Error::<T>::CollectionNotFound)?;
+
+        // Ensure origin is agent with custody and permissions for default portfolio.
+        let caller_portfolio =
+            Asset::<T>::ensure_agent_with_custody_and_perms(origin, ticker, portfolio_kind)?;
+
+        // Verifies if the NFT exists
+        ensure!(
+            PortfolioNFT::contains_key(&caller_portfolio, (&ticker, &nft_id)),
+            Error::<T>::NFTNotFound
+        );
+
+        // Burns the NFT
+        let new_balance = BalanceOf::get(&ticker, &caller_portfolio.did)
+            .checked_sub(ONE_UNIT)
+            .ok_or(Error::<T>::BalanceUnderflow)?;
+        BalanceOf::insert(&ticker, &caller_portfolio.did, new_balance);
+        PortfolioNFT::remove(&caller_portfolio, (&ticker, &nft_id));
+        MetadataValue::remove_prefix((&collection_id, &nft_id), None);
```

Comment:
Either we need to put a small limit on the number of metadata keys per NFT collection, or we need the caller to pass in a `max_metadata` or `metadata_count` parameter and use that in the weights.

Otherwise someone can create a NFT collection with thousands of metadata keys and mint thousands of NFTs.  Then burn them in large batches (utility.batch).  It would be expensive, but could still cause block production to go over the limit.

Reply:
As we discussed, since we already check the number of keys when creating the collection, I only changed the benchmark for the burn extrinsic.